[space-picker] First iteration

DmitrySharabin opened the pull request:
Partially addresses #123.

### Usage examples

```html
<space-picker></space-picker>
```

```html
<space-picker value="oklab"></space-picker>
```

```html
<space-picker spaces="oklch, p3, srgb"></space-picker>
```

```html
<space-picker spaces="oklch, p3, srgb" value="p3"></space-picker>
```

```html
<space-picker onspacechange="this.nextElementSibling.textContent = this.value"></space-picker>
<output></output>
```

LeaVerou commented on src/space-picker/space-picker.js:
```diff
@@ -0,0 +1,127 @@
+			default: () => Self.Color.spaces,
+			convert (value) {
+				// Drop non-existing spaces
+				let spaces = Object.fromEntries(Object.entries(value).filter(([id, space]) => space));
```
I don't think we should drop them, I think we should just not list them with a name. Just the id.

LeaVerou replied:
I.e. the name should default to the id.

DmitrySharabin replied:
Makes sense. Will fix it. Thanks!

LeaVerou commented on src/space-picker/space-picker.js:
```diff
@@ -0,0 +1,127 @@
+
+	static formAssociated = {
+		like: el => el._el.picker,
+		role: "combobox",
```
Doesn't this default to the element we specified in `like`? If not, it probably should.

DmitrySharabin replied:
[In our code, we try to default to `computedRole`](https://github.com/nudeui/element/blob/3c3297eed2506200998b43e29eaa0512c1f0bdcc/src/formAssociated.js/defineFormAssociated.js#L35). Unfortunately, it still isn't supported. It's probably under the flag in some browsers (not sure).

<img width="256" alt="image" src="https://github.com/user-attachments/assets/4ed1b55e-de8d-42e2-b124-6e10a328f6db">

LeaVerou replied:
We should just use `getComputedAccessibleNode()` instead. It's not like we don't have access to this information at all.

LeaVerou replied:
(in Nude Element, in case that wasn't clear)

DmitrySharabin replied:
Got it. Thanks. I'll send a PR with the change.

LeaVerou commented on src/space-picker/space-picker.js:
```diff
@@ -0,0 +1,127 @@
+		like: el => el._el.picker,
+		role: "combobox",
+		valueProp: "value",
+		changeEvent: "spacechange",
```
Shouldn't this use `change`?

LeaVerou commented on src/space-picker/space-picker.js:
```diff
@@ -0,0 +1,127 @@
+	};
+
+	static events = {
+		change: {
```
We should also retarget the `input` event.

LeaVerou commented on src/space-picker/space-picker.js:
```diff
@@ -0,0 +1,127 @@
+
+				value += "";
+
+				return Self.Color.Space.get(value);
```
This is again the same issue with the string/object props. We need a prop that contains your actual input as a string (`value`) and one that contains the color space object (likely `space` or `selectedSpace`).

LeaVerou commented on src/space-picker/space-picker.js:
```diff
@@ -0,0 +1,127 @@
+			this.value = this._el.picker.value;
+		}
+
+		this.dispatchEvent(new event.constructor(event.type, {...event}));
```
Doesn't defining the change event below take care of this?

DmitrySharabin commented:
Thank you so much for your review! Every comment is on point, and I'll try to address it ASAP.

DmitrySharabin commented:
@LeaVerou, I addressed all your feedback (thank you for it). Here are the changes I made:

- Use `value` for actual input (as a string) and `selectedSpace` for the corresponding space object
- Support custom grouping of color spaces via the `groups()` prop
- Don't drop non-existent spaces; replace them with the `{ id, name: id }` objects
- Expose the `valuechange` and `spacechange` events
- Retarget the `input` event
- In `formAssociated`, use `change` as the `changeEvent`

Here is an example of using a custom grouping (from your color palettes research):

```html
<space-picker id="space_picker" spaces="oklch, p3, srgb" value = "p3"></space-picker>

<script type="module">
	space_picker.groups = (spaces) => {
		let ret = {polar: {}, rectangular: {}};
		for (let id in spaces) {
			let space = spaces[id];

			if (space.id !== id) {
				continue;
			}

			let isPolar = space.coords.h?.type === "angle";
			ret[isPolar ? "polar" : "rectangular"][id] = space;
		}
		return ret;
	};
</script>
```

<img width="586" alt="image" src="https://github.com/user-attachments/assets/13b12bc7-8630-47bf-9f16-4fafd7b18fd2">


Could you please take a look at this PR one more time?

DmitrySharabin commented:
Thank you so much for your feedback! I tried to address all the issues you pointed out. Namely:

* Add the `groupBy()` prop that expects a space and returns the name of a group the space belongs to

Providing grouping looks like so now:

```html
<space-picker id="space_picker" spaces="oklch, p3, srgb" value = "p3"></space-picker>

<script type="module">
	space_picker.groupBy = (space) => {
		let isPolar = space.coords.h?.type === "angle";
		return isPolar? "Polar" : "Rectangular";
	};
</script>
```

<img width="589" alt="image" src="https://github.com/user-attachments/assets/32bb3adb-dd4a-466a-a130-b59f8bc8df39">


* Related to the previous one: `groups` is now a computed prop returning an object with all the groups
* Use different UI for cases with grouping and without

I hope I got everything right. If not, I'll be happy to fix it.

LeaVerou reviewed (approved):
Looks good! Please add a README, and add it to the list of upcoming components.

DmitrySharabin commented:
> Looks good! Please add a README, and add it to the list of upcoming components.

Thank you! Will do.

DmitrySharabin commented:
Done! It's alive and can be played with: https://elements.colorjs.io/src/space-picker/ 🎉
